libs/sdk-ui-kit/src/@ui/hooks/useListWithActionsKeyboardNavigation.ts

```diff
@@ -27,19 +32,66 @@ export function useListWithActionsKeyboardNavigation<Item, Action extends string
     actionHandlers,
     getItemAdditionalActions,
     isNestedList = false,
+    focusedIndex: focusedIndexProp,
 }: {
     items: Item[];
-    actionHandlers: { [key in Action | typeof SELECT_ITEM_ACTION]: (item: Item) => (() => void) | undefined };
+    actionHandlers: {
+        [key in Action | typeof SELECT_ITEM_ACTION]: (
+            item: Item,
+            e?: React.KeyboardEvent,
+        ) => (() => void) | undefined;
+    };
     getItemAdditionalActions: (item: Item) => Action[];
     isNestedList?: boolean;
+    focusedIndex?: number;
 }) {
-    const [focusedIndex, setFocusedIndex] = React.useState<number | undefined>(0);
+    const [focusedIndex, setFocusedIndex] = React.useState<number | undefined>(focusedIndexProp ?? 0);
     const focusedItem = focusedIndex === undefined ? undefined : items[focusedIndex];
 
     const [focusedAction, setFocusedAction] = React.useState<Action | typeof SELECT_ITEM_ACTION>(
         SELECT_ITEM_ACTION,
     );
 
+    // Store reference to the previously focused item to restore focus after loading
+    const previousFocusedItemRef = React.useRef<Item | undefined>(focusedItem);
+    const previousItemsLengthRef = React.useRef<number>(items.length);
+
+    React.useEffect(() => {
+        if (focusedIndexProp !== undefined) {
+            setFocusedIndex(focusedIndexProp);
+        }
+    }, [focusedIndexProp]);
+
+    // Handle loading state transitions
+    React.useEffect(() => {
+        const wasEmpty = previousItemsLengthRef.current === 0;
+        const isNowPopulated = items.length > 0;
+
+        if (wasEmpty && isNowPopulated && previousFocusedItemRef.current) {
+            const previousItemIndex = items.findIndex((item) => item === previousFocusedItemRef.current);
```

Comment:
I'm not sure if this should not be an argument provided to the hook. Predicate that finds the index. You can control the item selection logic in the component that uses the hook. I'm not sure if this approach when you use object identity will work everywhere. Usually, you will need ID compare logic or something similar.